File: problems/2025/07-17/decode-the-string/solution.py
def decode_string(s: str) -> str:
    num_stack = []
    str_stack = []
    curr = []
    num = 0
    for ch in s:
        if ch.isdigit():
            num = num * 10 + (ord(ch) - 48)
        elif ch == '[':
            num_stack.append(num)
            str_stack.append(''.join(curr))
            num = 0
            curr = []
        elif ch == ']':
            k = num_stack.pop()
            prev = str_stack.pop()
            curr = [prev + ''.join(curr) * k]
        else:
            curr.append(ch)
    return ''.join(curr)

if __name__ == "__main__":
    print(decode_string("3[b2[ca]]"))  # bcacabcacabcaca
    print(decode_string("3[ab]"))     # ababab

File: problems/2025/07-17/decode-the-string/solution.js
function decodeString(s) {
  const numStack = [];
  const strStack = [];
  let curr = '';
  let num = 0;
  for (const ch of s) {
    if (ch >= '0' && ch <= '9') {
      num = num * 10 + (ch.charCodeAt(0) - 48);
    } else if (ch === '[') {
      numStack.push(num);
      strStack.push(curr);
      num = 0;
      curr = '';
    } else if (ch === ']') {
      const k = numStack.pop();
      const prev = strStack.pop();
      curr = prev + curr.repeat(k);
    } else {
      curr += ch;
    }
  }
  return curr;
}

console.log(decodeString("3[b2[ca]]")); // bcacabcacabcaca
console.log(decodeString("3[ab]"));    // ababab
